flink-ml-lib/src/main/java/org/apache/flink/ml/clustering/kmeans/KMeansModel.java

```diff
@@ -0,0 +1,214 @@
+import org.apache.flink.runtime.state.StateInitializationContext;
+import org.apache.flink.streaming.api.datastream.DataStream;
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+import org.apache.flink.streaming.api.functions.sink.filesystem.bucketassigners.BasePathBucketAssigner;
+import org.apache.flink.streaming.api.functions.sink.filesystem.rollingpolicies.OnCheckpointRollingPolicy;
+import org.apache.flink.streaming.api.operators.AbstractStreamOperator;
+import org.apache.flink.streaming.api.operators.TwoInputStreamOperator;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+import org.apache.flink.table.api.Table;
+import org.apache.flink.table.api.bridge.java.StreamTableEnvironment;
+import org.apache.flink.table.api.internal.TableImpl;
+import org.apache.flink.types.Row;
+
+import org.apache.commons.collections.IteratorUtils;
+import org.apache.commons.lang3.ArrayUtils;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/** A Model which clusters data into k clusters using the model data computed by {@link KMeans}. */
+public class KMeansModel implements Model<KMeansModel>, KMeansModelParams<KMeansModel> {
+    private final Map<Param<?>, Object> paramMap = new HashMap<>();
+    private Table centroidsTable;
+
+    public KMeansModel() {
+        ParamUtils.initializeMapWithDefaultValues(paramMap, this);
+    }
+
+    @Override
+    public KMeansModel setModelData(Table... inputs) {
+        centroidsTable = inputs[0];
+        return this;
+    }
+
+    @Override
+    public Table[] getModelData() {
+        return new Table[] {centroidsTable};
+    }
+
+    @Override
+    public Table[] transform(Table... inputs) {
+        StreamTableEnvironment tEnv =
+                (StreamTableEnvironment) ((TableImpl) inputs[0]).getTableEnvironment();
+        DataStream<DenseVector[]> centroids =
+                tEnv.toDataStream(centroidsTable).map(row -> (DenseVector[]) row.getField("f0"));
+
+        String featureCol = getFeaturesCol();
+        String predictionCol = getPredictionCol();
+        DistanceMeasure distanceMeasure = DistanceMeasure.getInstance(getDistanceMeasure());
+
+        RowTypeInfo inputTypeInfo = TableUtils.getRowTypeInfo(inputs[0].getResolvedSchema());
+        RowTypeInfo outputTypeInfo =
+                new RowTypeInfo(
+                        ArrayUtils.addAll(inputTypeInfo.getFieldTypes(), Types.INT),
+                        ArrayUtils.addAll(inputTypeInfo.getFieldNames(), predictionCol));
+
+        DataStream<Row> input = tEnv.toDataStream(inputs[0]);
+        DataStream<Row> output =
+                input.connect(centroids.broadcast())
+                        .transform(
+                                "SelectNearestCentroid",
+                                outputTypeInfo,
+                                new SelectNearestCentroidOperator(featureCol, distanceMeasure));
```

Comment:
For the long run we may have an eager version of `SelectNearestCentroidOperator` that used for prediction.

Reply:
I agree. I will update KMeansModel to use BroadcastUtils to do the predication in an eager way after the bug in https://github.com/apache/flink-ml/pull/34 is fixed.